### PR TITLE
Mini-malloc option

### PR DESCRIPTION
**Pull request by kripken**
This adds a MALLOC option, where the user can pick which malloc/free implementation to use. The options are the existing dlmalloc, still the default, and the new emmalloc, which is about 1/3 the size. In small programs this can be very significant.

Performance-wise, emmalloc is small because it is very simple: all allocations are 8-byte, and there is no pooling, both of which are necessary to do well on huge amounts of tiny allocations. Because of that, is 35% slower on the havlak benchmark (which allocates tons of 12-24 size areas, an AST of small nodes) and 10% slower on lua-binarytrees (apparently each lua object is a separate allocation, and there's an AST of small nodes there too...). But on all other benchmarks the only noticeable difference is smaller code size.

In theory maybe we'd want to use the small allocator by default, or maybe just in say `-Os`. But for now it seems safest to stick with dlmalloc, and have the new allocator as an option.

This PR does two other noticeable changes:
 * Our sbrk used to align. The spec doesn't say it needs to, and it's redundant as dlmalloc and emmalloc check the alignment anyhow. So this PR removes that.
 * I noticed that building dlmalloc with NDEBUG for opt builds is better. So this PR makes dlmalloc a little smaller as well.

### CODE REVIEW

File: src/settings.js [hunk 1]
```diff
@@ -62,6 +62,15 @@ var TOTAL_STACK = 5*1024*1024; // The total stack size. There is no way to enlar
...
                                  // we need to copy the old heap into a new one in that case.
+var MALLOC = "dlmalloc"; // What malloc()/free() to use, out of
+                         //  * dlmalloc - a powerful general-purpose malloc
+                         //  * emmalloc - a simple and compact malloc designed for emscripten
+                         // dlmalloc is necessary for multithreading, split memory, and other special
+                         // modes, and will be used automatically in those cases.
+                         // In general, if you don't need one of those special modes, and if you don't
+                         // allocate very many small objects, you should use emmalloc since it's
+                         // smaller. Otherwise, if you do allocate many small objects, dlmalloc
+                         // is usually worth the extra size.
```
Comment (juj) [inline src/settings.js]: I wonder if an option `"none"` will be useful, to force user to supply implementation of `malloc` and `free` themselves? Not to say that should be implemented in this PR, but just wondering if that'll be feasible in this arch?

File: system/lib/emmalloc.cpp [hunk 2]
```diff
@@ -0,0 +1,1187 @@
...
+  // 4 which is 1 << 2, so the result should be 2
+  return 31 - __builtin_clz(x);
+}
+
+// Constants
+
+// All allocations are aligned to this value.
+static const size_t ALIGNMENT = 8;
+
+// Even allocating 1 byte incurs this much actual payload
```
Comment (juj) [inline system/lib/emmalloc.cpp]: Will allocating 0 bytes just allocate the 8 byte metadata, or also 8 bytes of payload?

File: system/lib/emmalloc.cpp [hunk 3]
```diff
@@ -0,0 +1,1187 @@
...
+  void setTotalSize(size_t x) { _totalSize = x; }
+  void incTotalSize(size_t x) { _totalSize += x; }
+  void decTotalSize(size_t x) { _totalSize -= x; }
+
+  size_t getUsed() { return _used; }
+  void setUsed(size_t x) { _used = x; }
+
+  Region*& prev() { return _prev; }
+  // The next region is not, as we compute it on the fly
+  Region* next() {
```
Comment (juj) [inline system/lib/emmalloc.cpp]: Should this function and all the above up to `getTotalSize` be `static`?

Comment (juj) [reply system/lib/emmalloc.cpp]: oh nm, first read that these were global scope functions, but they are instead members of `Region`.

File: tools/ports/binaryen.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import os, shutil, logging
 
-TAG = 'version_42'
+TAG = 'version_43'
```
Comment (juj) [inline tools/ports/binaryen.py]: Is this change intentional to coincide with this PR?

File: tools/system_libs.py [hunk 5]
```diff
@@ -293,19 +293,40 @@ def create_compiler_rt(libname):
...
+    elif shared.Settings.MALLOC == 'emmalloc':
+      base = 'emmalloc'
+    else:
+      raise Exception('malloc must be one of "emmalloc", "dlmalloc", see settings.js')
+    # only dlmalloc supports most modes
+    extra = ''
     if shared.Settings.USE_PTHREADS:
-      ret += '_threadsafe'
+      extra += '_threadsafe'
+      base = 'dlmalloc'
```
Comment (juj) [inline tools/system_libs.py]: This kind of validation ends up silently ignoring `-s MALLOC=emmalloc` directive when `-s USE_PTHREADS=1` and similar are set. Perhaps emcc.py could instead validate that `-s MALLOC=emmalloc` and `-s USE_PTHREADS=1` are mutually exclusive and so forth, and then here we could just pass-through `base = shared.Settings.MALLOC`?

**Comment by juj** [comment]
Very nice, looks like you went much farther in writing an allocator rather than doing the minimum. What kind of multithreading approaches can you think? Top level mutex guard, or guard each pool separately by a mutex? Linked list inserts and removes can be made lock free, though did not look really close to details if there's some data that inherently will need the mutexes.

lgtm, though probably may want to squash when merging?

File: src/settings.js (hunk 1 above)
Comment (dschuff) [reply src/settings.js]: I think regardless of what this setting is, the linking of malloc should behave as it does on native platforms; i.e. if you put your own implementation of `malloc` and `free` in an object file in the link, then that will be used instead of the emscripten-provided one.

Comment (juj) [reply src/settings.js]: Yeah, that's right, and that's something we do already test to work. This replacement however can be a bit of magic, since if something changes, e.g. one forgets to link to own malloc, there is generally no error message, and one silently gets the built-in malloc, so thought this kind of option would be nice to explicitly kill the original malloc from being considered, in which case one would get a nice missing symbol error instead if e.g. after build system refactoring one lacks linking in their own malloc.

File: system/lib/emmalloc.cpp [hunk 6]
```diff
@@ -0,0 +1,1187 @@
+/*
+ * Simple minimalistic but efficient malloc/free.
+ *
+ * Assumptions:
+ *
+ *  - 32-bit system.
```
[the remark this answers is not in the document]
Comment (dschuff) [reply system/lib/emmalloc.cpp]: Oh I didn't realize at first that this is a cpp file, we should totally add some static asserts to check these assumptions.

File: system/lib/emmalloc.cpp [hunk 7]
```diff
@@ -0,0 +1,1187 @@
...
+  // never want to modify it ourselves.
+  char* payload() { return &_payload[0]; }
+};
+
+// Region utilities
+
+static void* getPayload(Region* region) {
+  assert(((char*)&region->freeInfo()) - ((char*)region) == METADATA_SIZE);
+  assert(region->getUsed());
+  assert(sizeof(FreeInfo) == ALLOC_UNIT);
```
Comment (dschuff) [inline system/lib/emmalloc.cpp]: These can be static asserts and moved up to where these values are defined.

File: system/lib/emmalloc.cpp [hunk 8]
```diff
@@ -0,0 +1,1187 @@
+/*
+ * Simple minimalistic but efficient malloc/free.
+ *
+ * Assumptions:
+ *
+ *  - 32-bit system.
+ *  - Single-threaded.
```
Comment (dschuff) [inline system/lib/emmalloc.cpp]: I had thought there might be a way at compile-time to get the value of the compiler's thread-model (e.g. "single" or "posix") so we could assert on that too, but I didn't find one. It might be worth adding that to clang for wasm, since we expect users to care more about the distinction that on other platforms.

File: system/lib/emmalloc.cpp [hunk 9]
```diff
@@ -0,0 +1,1187 @@
...
+// 2^K bytes.
+//
+// Note that there is no freelist for 2^32, as that amount can
+// never be allocated.
+
+static const size_t MIN_FREELIST_INDEX = 3;  // 8 == ALLOC_UNIT
+static const size_t MAX_FREELIST_INDEX = 32; // uint32_t
+
+static FreeInfo* freeLists[MAX_FREELIST_INDEX] = {
+  NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
```
Comment (dschuff) [inline system/lib/emmalloc.cpp]: The nulls could all be nullptrs.

File: system/lib/emmalloc.cpp [hunk 10]
```diff
@@ -0,0 +1,1187 @@
...
+  // malloc() spec defines malloc(0) => NULL.
+  if (size == 0) return NULL;
+  // Look in the freelist first.
+  Region* region = tryFromFreeList(size);
+  if (!region) {
+    // Allocate some new memory otherwise.
+    region = newAllocation(size);
+    if (!region) {
+      // We failed to allocate, sadly.
+      return NULL;
```
Comment (dschuff) [inline system/lib/emmalloc.cpp]: I guess this allocator should respect ABORTING_MALLOC too?

File: src/settings.js (hunk 1 above)
Comment (kripken) [reply src/settings.js]: Yeah, we could add a "none" option that means it would be a link error to not provide your own malloc. I wonder if there isn't a more general way to do this. Not opposed, though.

File: system/lib/emmalloc.cpp (hunk 2 above)
Comment (kripken) [reply system/lib/emmalloc.cpp]: Allocating 0 just returns NULL, actually, which is allowed by the [docs](https://linux.die.net/man/3/malloc)
```
 If size is 0, then malloc() returns either NULL, or a unique pointer value that can later be successfully passed to free(). 
```

File: tools/ports/binaryen.py (hunk 4 above)
Comment (kripken) [reply tools/ports/binaryen.py]: Yeah, it depending on some fixes there. Meanwhile time has passed and a later version has been updated anyhow ;) So this will vanish in the merge.

File: tools/system_libs.py (hunk 5 above)
Comment (kripken) [reply tools/system_libs.py]: Good point, I'll make it show a clear error in those cases, instead of silently "fixing" things.

File: system/lib/emmalloc.cpp [hunk 11]
```diff
@@ -0,0 +1,1187 @@
+/*
+ * Simple minimalistic but efficient malloc/free.
+ *
+ * Assumptions:
+ *
+ *  - 32-bit system.
+ *  - Single-threaded.
+ *  - sbrk() is used, and nothing else.
+ *  - sbrk() will not be accessed by anyone else.
```
[the remark this answers is not in the document]
Comment (kripken) [reply system/lib/emmalloc.cpp]: It does assume `sbrk()` is not called by anyone else. Reading the docs, I don't see an official statement on it, but I see statements like "malloc will call sbrk(), and you shouldn't", so it seems already frowned upon. So many pointing this out here is redundant.

File: system/lib/emmalloc.cpp [hunk 12]
```diff
@@ -0,0 +1,1187 @@
...
+static const size_t METADATA_SIZE = ALLOC_UNIT;
+
+// How big a minimal region is.
+static const size_t MIN_REGION_SIZE = METADATA_SIZE + ALLOC_UNIT;
+
+// Constant utilities
+
+// Align a pointer, increasing it upwards as necessary
+static size_t alignUp(size_t ptr) {
+  return (size_t(ptr) + ALIGNMENT - 1) & -ALIGNMENT;
```
[the remark this answers is not in the document]
Comment (kripken) [reply system/lib/emmalloc.cpp]: Yes, thanks, fixing.

File: system/lib/emmalloc.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (kripken) [reply system/lib/emmalloc.cpp]: I'll clarify, and add asserts - this about pointers being 32 bit.

File: system/lib/emmalloc.cpp (hunk 7 above)
Comment (kripken) [reply system/lib/emmalloc.cpp]: Fixing.

File: system/lib/emmalloc.cpp [hunk 13]
```diff
@@ -0,0 +1,1187 @@
...
+
+static bool isPowerOf2(size_t x) {
+  return __builtin_popcount(x) == 1;
+}
+
+static size_t lowerBoundPowerOf2(size_t x) {
+  if (x == 0) return 1;
+  // e.g. 5 is 0..0101, so clz is 29, and we want
+  // 4 which is 1 << 2, so the result should be 2
+  return 31 - __builtin_clz(x);
```
[the remark this answers is not in the document]
Comment (kripken) [reply system/lib/emmalloc.cpp]: I suppose it could be `sizeof(size_t) * 8 - 1` but that seems less nice...

File: system/lib/emmalloc.cpp (hunk 10 above)
Comment (kripken) [reply system/lib/emmalloc.cpp]: ABORTING_MALLOC should just work, that is handled by sbrk (it will abort when it can't grow memory).

**Comment by kripken** [comment]
@juj "What kind of multithreading approaches can you think?"

I don't have much thoughts there yet. Maybe this will end up mostly interesting for the single-threaded case anyhow.

**Comment by kripken** [comment]
Merged current incoming and addressed review comments, thanks!

File: system/lib/emmalloc.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (dschuff) [reply system/lib/emmalloc.cpp]: Could do something like `#define SIZE_BIT sizeof(size_t) * CHAR_BIT` above where we check all of the assumptions.

[the remark this answers is not in the document]
Comment (kripken) [reply system/lib/emmalloc.cpp]: Cool, done.

**Comment by juj** [comment]
Sweet! lgtm

**Comment by kripken** [comment]
@sbc100 sorry about that, turns out it's easy to fix though, see #6316
